docs/site/Life-cycle.md

```diff
@@ -14,6 +14,75 @@ to control the transition of states of `Application`.
 - start(): Start the application
 - stop(): Stop the application
 
+## Application states
+
+The initial state of an application is `created` when it's instantiated. The
+state can transition as follows by `start` and `stop`:
+
+1. start()
+
+- created -> starting -> started
+- stopped -> starting -> started
+- started -> started (no-op)
+
+The `start` throws an error if it's called with an invalid application state,
+such as `starting` and `stopping`.
```

Comment:
Should we allow `start` as a no-op while the application is in a `starting` state?

Reply:
IMO, no. Consider the following code:

```ts
await app.start(); // The promise should be not resolved until the application is started
// We assume the app is now in `started` state
```

The other option is to have `app.start()` wait for `started` event while it's starting.

Reply:
> The other option is to have `app.start()` wait for `started` event while it's starting.

That's what I would expect from `app.start()` method. As a developer calling `app.start`, I want to ensure the app is running. I don't necessarily care if it's stopped or started, I want to get it to the started state.


In my experience, APIs that offer [idempotence](https://en.wikipedia.org/wiki/Idempotence) are easier to use.